### ISSUE TITLE
ConfigProvider contextuality is wrong

### ISSUE THREAD

**Issue by rmannibucau** [eclipse/microprofile-config#13]
ConfigProvider has 1 big issues in the current design:

-> it handles the lookup which is wrong, in particular in the microprofile context which has CDI has a strong backbone

Also note that - assuming it is a valid design - it assumes there is a context which is a classloader which is wrong in several java applications so key should be generic probably (interface with only hashCode() and equals()?)

Concretely the config API should just allow to:
1. build a config (with sources, converters, accessors, autoloaded from SPI or manually built for libraries for instance)
2. provide a default config instance in CDI context (@ApplicationScoped)
3. if considered relevant provide a bootstrap config (the default one) to let configure bootstrap components like CDI extensions (this can require some CDI work to ensure it is not hacky but technically there is no blocker, in particular in microprofile context)

In short: there is no point but drawbacks to re-implement what CDI already does which is the contextuality of an instance (Config there).

To make it more obvious: this issue is about removing any contextuality/lookup feature from the config project and let it be part of the container/IoC integration only.

**Comment by Emily-Jiang** [comment]
The CDI support will come later in the next release. By the way, the config instance needs to be provided by a producer as a config instance is per classloader. By the way, the Config design is trying to align with Config JSR proposal and we will put the CDI integration shortly after the 1.0 release. We had a long discussion on the thread [https://groups.google.com/forum/#!topic/microprofile/VJEEAOsVj5E[126-150]](url)

**Comment by rmannibucau** [comment]
@Emily-Jiang no, the producer is not needed? At least there is no constraint at all to require it. Also the "per classloader" rule is wrong and depends the environment. The Config JSR as well doesn't need any contextuality since the IoC or container (OSGi for instance) will handle it.

So concretely there is not a single reason to implement such logic in a config solution, and there is no point having a classloader based or lookup solution in a config solution.

Please revise these statements which are not backed by anything concrete.

**Comment by struberg** [comment]
In my spec document and API JavaDoc I only say 'per Application'.

It's totally up to the implementation whether it uses a ThreadLocal (as in your sample), a WeakMap<ClassLoader, ..>, an OSGi BundleActivator, or even a CDI Extension (also automatically 1 per Application). 

The API does _not_ mandate to use the ClassLoader (except to 'identify' the Application). All the rest is a detail of the underlying implementation and should not affect the API.

**Comment by rmannibucau** [comment]
And CDI is per "X" already so we are able to reuse this X with microprofile stack. A spring app would just register a config in its context, etc...

No sense to create another context IMHO, can only lead to "I got the wrong config"

**Comment by Emily-Jiang** [comment]
@rmannibucau for an example, if a class in war1, it can see the config inside the war1 while a class in war2, it can see the config inside the war2 but not war1's config packaged inside the war. As for the OSGi bundle, the whole bundle will see the same config.

**Comment by rmannibucau** [comment]
@Emily-Jiang if you have war1 and war2, CDI context will be used. If the container doesn't handle isolation properly then the container has a bug but it is not a reason to reimplement another light IoC in the config, isn't it? In consistent containers you'll get 2 CDI contexts (1 per webapp) and the config will inherit what it needs from the upper level (ear lib + container if needed) and have specific sources/converters per webapp. It still sounds sane to me. Am I missing something?

**Comment by Emily-Jiang** [comment]
@rmannibucau CDI integration is not in the current design. It will come later.

**Comment by rmannibucau** [comment]
@Emily-Jiang was not the point, the point is the config should just allow to build a Config instance, period. Then it will be nice to integrate it with CDI, Spring etc and all the demonstration was just about showing there is no blocker. But the ticket is about *removing* from the config "spec" the contextuality the config doesn't need and doesn't even own.

**Comment by Emily-Jiang** [comment]
@rmannibucau misunderstood what you meant. Do you mean the ConfigProvider should not expose the following method:
public static Config getConfig(ClassLoader cl)?

In OSGi, it should go through CDI?

**Comment by rmannibucau** [comment]
Ok, let's step back.

ConfigProvider is a "spec" trick to find an implementation of Config but in current form it also holds the lookup logic to find the config instance (getConfig()*). All these methods are actually useless.

You should be able to build a config (pseudo code so don't shout on the naming please ;)):

    Config config = Config.start().useDefaults().addSource(new MySource()).....create();

Then what is done with this ("detached from anything") instance is up to the integration.

CDI will provide it in extension (through events then through a bean for runtime for instance - just one cdi proposal but it is out of topic there),
Spring will add it as a bean,
Plain SE applications will use as in any plain SE applications setters or constructors to pass the configuration instance
etc...

So concretely the configuration location is completly out of topic in there and belongs to other projects/specs. Therefore I think it is important to clean up the API and remove anything related to such logic.

**Comment by struberg** [comment]
In OSGi the ClassLoader usually also implements the org.osgi.framework.Bundle interface. Or at least it's always possible to resolve the current Bundle if you have the ClassLoader. 
This is also very usable for EAR environments where the current TCCL might not always reflect the proper part of the application.

@rmannibucau Yes the ConfigProvider knows how to resolve the Config. But nowhere in the API nor in the spec pdf is stated that it is 1:1 with the ClassLoader! It says "Application". 
Any implementation can freely decide how it resolves the current Config of the Application. May that be through a BundleActivator in OSGI, CDI or a Map<ClassLoader, Config>. I totally don't care, all that is possible.

**Comment by rmannibucau** [comment]
@struberg several framework integrating with OSGi will implement a classloader on the fly without that Bundle so you need a bundle locator so it doesnt help much.

Config.getConfig(Classloader) is quite explicit and broken by design. That said my proposal is to drop static utility to find the config. It doesn't remove any use case in term of usage and clean up the API preventing cases where it will not work (trivial to find with OSGi, embedded servers at least).

> Any implementation can freely decide

Also means you forget user from the starting design point of view. A user doesnt want to implement a spec but just use it directly without having to implement a SPI or an extension before.

So let's cover it in order:

1. how to build a config instance
2. how to integrate with CDI
3. if needed how to provide a ConfigLocator - the getConfig* - on top of Config if required (but as explained it is not I think)

**Comment by struberg** [comment]
> A user doesnt want to implement a spec but just use it directly 
> without having to implement a SPI or an extension before.

Exactly! And no line more.

That's what we provide with this proposal. A user can simply user ConfigProvider.getConfig() and he will get the auto-discovered Config out of the box.

Of course we also added a few additional methods to manually build a config and register it since it seems there is demand (you and Gunnar voted for it).

**Comment by rmannibucau** [comment]
@Inject Config config;

for CDI and spring is 1 line
1. without lookup issues (ConfigProvider.getConfig() is easy to break)
2. is natural compared to an old singleton like API which replaces the IoC for no real reason in "modern" applications

**Comment by Emily-Jiang** [comment]
@rmannibucau In order to build a config instance without CDI, we use this builder pattern. When we put CDI in and cdi is enabled, the programmatic lookup via ConfigProvider.getConfig() is not used.

**Comment by rmannibucau** [comment]
@Emily-Jiang not even using CDI lookup? it is worse than I thought then, you have a dead API, sometimes working, sometimes not.

you don't need it in any case, just ensure a Config instance can be built with auto or not discovery, then add it in the application context (whatever it means depending the application: spring context, cdi context, OSGi services etc...)

**Comment by struberg** [comment]
If you read my version of the proposal then you can see that I already suggest this
http://long4.me/javaconfig/microprofile-config-spec.pdf

See the DynamicConfigUsageSample class.

We could even go further to support injecting an owner-style config like Gunnar suggested. But I'd rather leave this for the next iteration.

**Comment by rmannibucau** [comment]
No Mark, I'm *not* (yet) about the CDI API, I'm about removing any logic linked to a locator from the config core spec. We agree on the final API but I showed that we can cut some methods from the core spec without affecting users and usages. So I think we should do it to avoid to release something we loose control on later.

**Comment by Emily-Jiang** [comment]
Maybe for the first version, we leave out the setConfig and releaseConfig. As @rmannibucau say, in the next release, we have CDI to manage its lifecycle. The setConfig and releaseConfig are dead. Basically, the config we build now is unmanaged, no caching. Thoughts?

**Comment by rmannibucau** [comment]
if provider creates a config it needs to release it (can be hidden in Config#close() BTW but depends your style preferences I guess) for symmetry. If CDI you will create and release it, for spring also, etc...

What is NOT needed is to manage any context so getConfig() and getConfig(WhateverKeyAndInParticularClassLoader) should be dropped for v1 at least.

**Comment by struberg** [comment]
The Config will automatically get release when the application is stopped. 

And of course, CDI can manage the start and end of the Config Lifecycle. But we do not always have CDI. For example JPA doesn't know whether it will be used in conjunction with CDI or Spring. 

And the other problem is that such a config would only be usable after the AfterDeploymentValidation. You would need a complex system of events inside CDI Extensions to make it at least working in CDI Extensions. But there is no way to e.g. use it in JPA, Bean Validation etc.

What you would do is to trade in a ConfigSource which gets picked up via ServiceLoader with a CDI Extension which AGAIN gets picked p via ServiceLoader. So you would just unnecessarily add complexity without gaining much. We would actually loose flexibility. I see no reason to castrate ourselves.

**Comment by rmannibucau** [comment]
JPA: whatever you use CDI or Spring you will create the unit and therefore be able to pass the config there (like Validator or BeanManager in the spec) so not an issue

The gain is to ensure it works where today this API prevents it to work in several common java env (I don't even speak of leaks). Will also allow to rely on container feature and guarantee sources are singletons (you can't otherwise and it can be issues with actual source implementations on relying on external resources).

Not sure i get your point on AfterDeploymentValidation, it is doable with CDI even if less trivial than with some other IoC and JPA/BVal will be excluded with this as well before AfterDeploymentValidation so not sure where it leads.

**Comment by rmannibucau** [comment]
Another way to understand this issue is this one:

What is the role of ConfigProvider? build a Config instance based on an implementation (XImpl), if it needs to handle the lookup you will get XOSGiImpl, XEEImpl, XCDIImpl, XTomEEImpl, XWildflyImpl etc...

so you will extract that in LocatorProvider to be able to lookup a Map<Key, Service> wherever you are

then you realize this LocatorProvider is the IoC and therefore you don't need this layer in this spec

**Comment by Emily-Jiang** [comment]
I think this further and I agree with @rmannibucau that we should delete the setConfig and clearConfig. All programmatically created config is unmanaged. All config created by CDI are managed.

**Comment by struberg** [comment]
We come a full circle back to my original proposal:

- getConfig() -> autodiscovery and managed 'Singleton'
- newConfig() -> ConfigBuilder for creating a fresh and unmanaged Config each time

**Comment by Emily-Jiang** [comment]
no release and register methods should be there.

**Comment by struberg** [comment]
You remember that we added them only after the hangout due to popular requests from @gunnarmorling, @heiko-braun and others?

The release() method is btw needed to prevent mem leaks in certain environments (IF we manage the instances ourselves).

**Comment by rmannibucau** [comment]
Yes close() is needed. In term of API it can be Config#close() or Provider#close(Config) (I prefer the first one personally but both work).

The question is more: do we get a managed config or not. I think it 100% belongs to integrations so raw spec should not have it in the API and dedicate that part to CDI ("a default Config bean is automatically added....").

makes sense?

**Comment by heiko-braun** [comment]
I think I agree with @rmannibucau about the contextuality issue.  The appearance of classloader terms in the high level API is something I stumbled across also. IMO the API should not make any assumptions about this and hence I would suggest we revisit `Config.getConfig(ClassLoader cl)` and `Config.getConfig()` methods.

Without, the API for accessing configuration would be left to `ConfigBuilder.getBuilder()`, which I think in managed environments (i.e. with CDI) would be the common approach anyway and thus the removal of `Config.getConfig()` would not impose any suability drawbacks. 

The only cases where I see the usage of `Config.getConfig()` being useful and reasonable are non-managed environments like Java SE applications. But I think here each config api implementor is free to supply auxiliary extension to provide this level of convenience.

**Comment by struberg** [comment]
@heiko-braun @rmannibucau Please note that most of the 'for ClassLoader' got replaced with 'for the Application as indicated by the ClassLoader'.

So the spec would basically require "1 config per Application". Whatever your 'Application' terminus is. This would work well for WARs, standalone JARs, OSGi and even EARs. 

The current version of the API still misses my cleanup though. That might have gone missing with one of the PRs which got rejected. I'll fix that.
This is a fine (but important) notion which went missing in comparison to my original proposal
https://github.com/struberg/javaConfig/blob/master/api/src/main/java/io/microprofile/config/ConfigProvider.java#L75

**Comment by heiko-braun** [comment]
> So the spec would basically require "1 config per Application". Whatever your 'Application' terminus is. This would work well for WARs, standalone JARs, OSGi and even EARs.

I fully agree with need for isolation, but I think this is more of a cross-functional requirement, that commonly can be thought of "as being identified via a ClassLoader". But "commonly" doesn't mean this statement holds in all environments.

**Comment by rmannibucau** [comment]
Rephrased: config is just one component of an application like a Validator or an EntityManagerFactory, its contextuality is only relevant when started to load default instances but then it is a "detached" instance the application writer wire up to its need. There are tons of solutions for that, CDI for EE, Spring/Guice/Pico/Dagger/OSGi registry/... for others so config doesn't need any contextuality in its code.

Integration can provide a default instance which would be contextual but that's out of scope for the core API.

**Comment by Emily-Jiang** [comment]
@heiko-braun ConfigProvider.getConfig() is very useful in the auto discovery scenarios as you don't need to use a builder and build all config sources again. 

@rmannibucau you are talking about the CDI integration. @Inject Config myConfig will give an instance just right for you. However, the api/spi is for programmatical look up and its contextual needs to be manged. e.g. The config properties in war1 should not be accessible for war2.

**Comment by rmannibucau** [comment]
@Emily-Jiang :

1. ConfigProvider.getConfig()  which assumed a contextuality owns by config impl is likely worse than (pseudo code) ConfigProvider.newBuilder().create() which is still using the auto discovery
2. "the api/spi is for programmatical look up and its contextual needs to be manged": just explained this need doesn't belong to config but the container it runs in, exactly like a Validator instance or EntityManagerFactory. Does bean validation povides a Validator.getCurrentInstance() or EntityManagerFactory.getCurrentInstance()? No, cause this contextuality belongs to CDI or EE (this last point being historical and pre-CDI). Same in spring: you don't use a bval or jpa lookup but a spring context lookup. In summary there is no point having such a config since it belongs to the IoC layer of an application/library and it is done outside the config spec anyway. This is the point :). I just explained that to be smooth a default CDI integration should be provided - even if not mandatory. Is it clearer this way?

**Comment by gunnarmorling** [comment]
The essential question to me is do we assume CDI as a basis or not. If yes, then `ConfigProvider` indeed is more or less useless (besides exposing a way for obtaining a builder), as users would simply obtain `Config` instances via `@Inject`.

@rmannibucau so what exactly do you suggest? Which methods should go away?

**Comment by Emily-Jiang** [comment]
@rmannibucau Bean Validation provides a Validator instance. Similarly Config needs to provide a Config instance.

**Comment by rmannibucau** [comment]
@gunnarmorling I wouldn't assume CDI is there in the api but keep a default config instance if CDI is there (no blocker to do it with an extension). Concretely all the API should only be able to create a ready to use Config instance so:

1. Config getConfig()
2. Config getConfig(ClassLoader cl)
3. setConfig(Config config, ClassLoader classLoader)

should be dropped

**Comment by rmannibucau** [comment]
@Emily-Jiang bean validation doesn't manage any context, CDI does on *top* of bean validation, not the opposite (which is current proposal in config)

**Comment by Emily-Jiang** [comment]
@rmannibucau We are supporting both CDI approach and non-CDI approach.
The methods you listed is for all CDI approach but for non-CDI approach.

**Comment by rmannibucau** [comment]
@Emily-Jiang assuming it is true then you are broken in CDI environment and even more in CDI + non CDI applications (partly CDI and partly spring are common) so this API is wrong anyway. But also note that we don't need them without loosing any feature and even we encourage a correct design and not the user to rely on a wrong contextuality without noticing. If you miss some contextuality in a plain standalone (note plain standalone often have its own container but lets assume not for the exercise) then it is trivial to import pico container or any light container (even a map) to replace these methods removing all the ambiguity it can bring to real applications

**Comment by gunnarmorling** [comment]
@rmannibucau agreed, but BV also provides you with a way (validation.xml) to customize the config of the "default validator factory", i.e. the one you obtain from `Validation#buildDefaultValidatorFactory()` and which also is the one exposed via CDI etc.

The methods `ConfigProvider#setConfig()` and `releaseConfig()` were meant to allow for such customization of what's returned as the "default config", as no other agreement could be reached on how the default config could be configured.

Of course we could leave that to specific implementors for the time being.

If you say you don't assume CDI as a basis but also want to get rid of `getConfig()`, then how would one obtain a config at all? Seems it means that all integrators need to implement the correct builder invocation to match whatever is defined by mP Config as the default sources. Or are you saying you'd leave that out, too?

**Comment by rmannibucau** [comment]
@gunnarmorling all this customization logic belongs to the CDI integration (so outside ConfigProvider anyway) and idea was to fire in an extension the builder to let customize it from other extensions which works well and doesn't require any new API.

For the getConfig() part:

- in CDI I think we all agree we'll do @Inject Config config;
- in Spring: will be the same (@Autowired if you are nostalgic ;))
- in Guice: same again
- in plain standalone: what do we use in standalone? setters/constructors? then it will not change

So concretely we don't miss anything. In term of libraries we'll get as already the case for all spec a CDI integration, a Spring integration to lookup the Config instance etc.. but there is no blocker to drop these methods.

Side note: I didn't say we'll drop releaseConfig() but it will be stateless (or we move it in Config#close() which would work and would be closer to bean validation/jpa/user habits - but it is 1-1 in term of complexity/feature)

**Comment by Emily-Jiang** [comment]
@gunnarmorling agreed.
@rmannibucau I don't understand why it is broken. I see them as a short hand route to get a config. Consider the 3 use cases from Mark.
1.) auto-discovery 
2.) unmanaged new Config each time you use a ConfigBuilder 
3.) managed Config as replacement for 1.) if someone wants to manually create a Config.
For 1) you just use ConfigProvider.getConfig(); 2, 3) ConfigProvider.getBuilder().addDefaultSources().withSources(...).build(). For 3) you can then call ConfigProvider.getConfig() to retrieve the config you built earlier.

**Comment by rmannibucau** [comment]
@Emily-Jiang it is broken cause enforce all environments to create a new implementation vs letting all environment - already existing - contextuality handling that concern. So instead of having a "IBM"/"ReadHat"/"Apache"/"Cook guy on github" implementations you'll get an "IBM OSGi"/"IBM EE"/"IBM Standalone" etc... implementations which are all conflicting and prevents a bunch of use cases. Also nothing guarantee you your default contextiality works (Map<ClassLoader, BeanManager> was often broken and replaced by CDI.current() with success so why breaking a new spec the same way old libraries were?). With my proposal of moving the default instance to CDI integration all the use cases are still managed, the API get rid of the contextuality making all frameworks happy and letting all config implementations not caring of something they don't control at all.

**Comment by heiko-braun** [comment]
> So instead of having a "IBM"/"ReadHat"/"Apache"/"Cook guy on github" implementations you'll get an "IBM OSGi"/"IBM EE"/"IBM Standalone" etc

I think what @rmannibucau refers is what I said above as well: You cannot guarantee the the assumption behind `ConfigProvider.get()` hold in all environments.

The other difficulty is reflected in your previous statement:
> The methods you listed is for all CDI approach but for non-CDI approach.

Having these two in the same API means, that depending on your context, one will not work as expected.

**Comment by gunnarmorling** [comment]
The issue I see with what you propose @rmannibucau is that all these integrators need to put the default config sources in into place, in the right order. Let's say we define that by default system props, environment variables and java-config.properties should be enabled, all the integrators (CDI, Spring etc. but also standalone users such as other 3rd party libraries) need to make sure they obey to this order. For sure it's doable, but it puts the onus to do so on each and every integrator instead of providing it in place in the config API itself.

Also it requires a way to actually obtain config sources for system props, environment variables and so on. Last time I checked there was nothing for that in the API (i.e. no public `ConfigSource` implementations and also no public factory methods to obtain them).

**Comment by rmannibucau** [comment]
@gunnarmorling think there is a misinterpretation there. I don't want to drop a default loading, i want to drop the fact the instance is tracked:

    builder.build()

vs

    builder.ignoreDefaultLoading().addSource(new MySource()).build()

then the contextuality is up to the environment/user

You have a good point about having some default impl for sources but this is not directly linked to that ticket

**Comment by gunnarmorling** [comment]
Ah, so you are saying that the builder would apply the default sources
unless one opts out ("ignoreDefaultLoading()"). That should work I guess.

2017-02-07 14:28 GMT+01:00 Romain Manni-Bucau <notifications@github.com>:

**Comment by rmannibucau** [comment]
Yes - just realized it was pretty close to bean validation again but that's the idea.

**Comment by gunnarmorling** [comment]
The only thing missing now is a way to influence that sequence of default
sources, e.g. to put an external config server source in there. Likely I
cannot go and add it to all usages of getBuilder() (which e.g. may be in a
library I cannot alter).

I had suggested a meta-config file (e.g. java-config.xml, akin to BV's
validation.xml) for this, listing all the sources of the default config.
But we couldn't agree on this, Mark had some arguments against it and
proposed the auto-discovery of sources present in the environment.

This really is how ConfigProvider#setConfig() came to be, it'd allow to set
what's returned as the default config.

**Comment by rmannibucau** [comment]
@gunnarmorling Mark has a good proposal based on CDI extensions which solves it for CDI, in Spring/Guice I guess we don't have the issue since we init the instance in the code/xml. Only thing we can maybe take care is to use a fluent setter instead of just semantic method (ignoreX() vs setIgnoreX()) to ensure common factories can work with it.

cc @struberg : do you have the link to the extension solution?

**Comment by struberg** [comment]
Can we do a hangout to discuss this through? I currently fail to see the problems.

**Comment by Emily-Jiang** [comment]
I was going to say the same, Mark! I think we are going in circles. I'll schedule a hangout soon.

**Comment by struberg** [comment]
we did have this hangout 3 weeks ago and agreed that while not perfect it works 'good enough'.
